XIPs/xip-31-message-history.md

```diff
@@ -0,0 +1,435 @@
+
+### Remote Message Backups
+
+Remote message backups work by having Backup Requesters send a special message type (`MessageHistoryBackupRequest`) across the XMTP network to all other installations signed by the same blockchain account as the sender (Message Backup Providers).
+
+Upon receipt of these messages, Message Backup Providers should display a prompt to the user asking whether they consent to share their message history with the requesting installation. The prompt should display the Verification PIN from the request, which will also be displayed in the Backup Requester's app to ensure the user approves the correct request. Some apps may choose to force the user to enter the PIN rather than display it as an additional layer of security.
+
+Upon approval, the app will convert its local database into a standard Message History Backup File, encrypt it, and upload it to the Backup Storage Provider specified in the `MessageHistoryBackupRequest`.
+
+For mobile apps already handling push notifications, `MessageHistoryBackupRequest`s would become a special case of push notification handling.
+
+#### Flow for Backup Requesters
+
+1. Get a list of all other installations associated with their blockchain account. Ignore any installations that have been revoked
+2. Generate a `MessageHistoryBackupRequest` for each installation (see below for Protobuf spec) and store in the local database
+3. Send the `MessageHistoryBackupRequest` messages to the normal inbound messaging topic for each installation
+4. Wait for a response from any of the Message Backup Providers
+5. For each `MessageHistoryBackupResponse`
+
+   5a. Ensure there is a `MessageHistoryBackupRequest` with a matching `requestId` stored in the database. If not, ignore.
+
+   5b. Ensure the `backupUrl` is on the same host as the requested `backupStorageProviderUploadUrl`. If not, ignore.
+
+   5c. Download the file from the `backupUrl` and decrypt using the credentials provided in the `MessageHistoryBackupResponse`. If the hash of the downloaded file does not match the hash in the `MessageHistoryBackupResponse`, abort.
+
+   5d. Load each message into the local database, ignoring any duplicate messages.
+
+   5e. Delete the `MessageHistoryBackupResponse` and all associated credentials.
+
+   5f. Set the status of the `MessageHistoryBackupRequest` to `Applied`.
+
+#### Flow for Message Backup Providers
+
+1. Receive a `MessageHistoryBackupRequest` as part of the normal message receiving flow.
+1. Retrieve the contact for the installation that sent the message and ensure it has not been revoked. If revoked, ignore.
+1. Ensure the installation that sent the message has a contact signed by the same blockchain account as the current user. If not, ignore.
+1. Convert all messages in the local database into a Message Backup File (maybe we want to chunk here?)
```

Comment:
OK to remove the question? Has this been resolved?

Reply:
yes, OK to remove the question